File: ios/Sources/BackgroundGeolocationPlugin/Domain/Journey.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 gachlab

import Foundation

/// A completed driving trip — a pure, immutable value object for the domain layer
/// (ARCHITECTURE.md domain set), the Swift twin of the Android `domain/Journey`. Where
/// `Trip` is the journey in progress, Journey is its finished aggregate: the distance
/// travelled, the span, and the computed `score`.
///
/// It was three loose arguments threaded through `detectorOnTripEnd`
/// (distanceMeters / durationMs / score); bundling them gives the detector and the bridge
/// a single value to carry to the `tripEnd` event. `startedAtMs`/`endedAtMs` are Unix
/// epoch **milliseconds** — same unit and field names as the Android twin, so the
/// timestamps are safe to emit cross-platform (avoids a seconds-vs-ms trap).
struct Journey {
    let id: String
    let startedAtMs: Int64
    let endedAtMs: Int64
    let distanceMeters: Double
    let score: TripScore

    /// Trip span in milliseconds.
    var durationMs: Int64 { endedAtMs - startedAtMs }

    /// The Journey that `trip` becomes when it ends at `endedAt` (seconds) with `score`.
    static func completed(_ trip: Trip, endedAt: Double, score: TripScore) -> Journey {
        Journey(id: trip.id,
                startedAtMs: Int64(trip.startedAt * 1000),
                endedAtMs: Int64(endedAt * 1000),
                distanceMeters: trip.distanceMeters, score: score)
    }
}

File: android/src/main/java/com/gachlab/geolocation/domain/Journey.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 gachlab

package com.gachlab.geolocation.domain

import com.gachlab.geolocation.TripScore

/**
 * A completed driving trip — a pure, immutable value object for the `domain/` layer
 * (ARCHITECTURE.md domain set). Where [Trip] is the journey in progress, Journey is its
 * finished aggregate: the distance travelled, the span, and the computed [score].
 *
 * It was previously three loose arguments threaded through `onTripEnd`
 * (distanceMeters / durationMs / score); bundling them gives the hub and the bridge a
 * single value to carry from the detector out to the `tripEnd` event.
 */
data class Journey(
    val id: String,
    val startedAtMs: Long,
    val endedAtMs: Long,
    val distanceMeters: Double,
    val score: TripScore,
) {
    /** Trip span in milliseconds. */
    val durationMs: Long get() = endedAtMs - startedAtMs

    companion object {
        /** The Journey that [trip] becomes when it ends at [endedAtMs] with [score]. */
        fun completed(trip: Trip, endedAtMs: Long, score: TripScore): Journey =
            Journey(trip.id, trip.startedAtMs, endedAtMs, trip.distanceMeters, score)
    }
}
